### PR TITLE
CONTRIBUTING.md: "no squash merges"

### PR DESCRIPTION
**Pull request by ncfavier**
@Infinisil first explained this to me in https://github.com/NixOS/nixpkgs/pull/104836#issuecomment-736123877 and there is evidence of at least one other person being confused by this, so let's make it explicit.

Please feel free to suggest wording improvements.

### CODE REVIEW

**Comment by mweinelt** [comment]
I've used squash merge on a few occasions, when communication simply did not seem to work. I have certain PRs on my mind, but will refrain from playing the blame game here.

**Comment by Artturin** [comment]
Squash merges can be disabled in the repo settings
Screenshot_20221015_191237

**Comment by jonringer** [comment]
If we really want to enforce it, I think @Artturin suggestion of just having an admin disable the possibility makes the most sense.

**Comment by infinisil** [comment]
> I've used squash merge on a few occasions, when communication simply did not seem to work. I have certain PRs on my mind, but will refrain from playing the blame game here.
> 
> I wish this would work.

If the author didn't disable the ability for contributors to push to their branch, you can do the rebasing for them. The only problem is that commit signatures won't be persisted then.

**Comment by ncfavier** [comment]
I think this discussion is orthogonal to the proposed addition, since it only says that "we prefer to use" squash merges?

**Comment by mweinelt** [comment]
I think I'm with @jonringer here. If we don't want it let's agree to disable it.

**Comment by jonringer** [comment]
> Sure, but that does not contradict the need for documenting our policy.

Sure, documentation is fine, especially for stating WHY we don't want squashes, but disabling squash merges just allows for the "impossible state to no be represented"

**Comment by mweinelt** [comment]
Squash merged just now in https://github.com/NixOS/nixpkgs/pull/201190 to integrate an accepted review change. Sometimes squashing is just easier.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -53,6 +53,10 @@ In addition to writing properly formatted commit messages, it's important to inc
 
 For package version upgrades and such a one-line commit message is usually sufficient.
 
+We prefer not to use the "squash merge" feature in nixpkgs: in order to keep as much information as possible in the commit history, we expect pull requests to consist of self-contained commits as described above.
```
Comment (SuperSandro2000) [inline CONTRIBUTING.md:56]: What do you mean by ``to keep as much information as possible in the commit history``? The improved [squash merge by Refined GitHub](https://github.com/refined-github/refined-github/wiki/Extended-feature-descriptions#clear-pr-merge-commit-message) for example pastes all the commit titles into the squash merge body. The body would be lost but PRs that have well written commit messages are usually not squash merged.

Also we do not expect commits to be fully self contained. It is totally fine to also update meta.homepage, cleanup up an old ToDos or improve formatting in an update commit. Also nodePackages and vimPlugins do auto commits for many packages at once by design and are not really self-contained.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -53,6 +53,10 @@ In addition to writing properly formatted commit messages, it's important to inc
...
+This means that, after addressing review comments and before the PR is merged, you will sometimes need to rewrite your branch's history and then force-push it with `git push --force-with-lease`.
+Useful commands to be comfortable with are `git commit --amend`, `git commit --fixup` and `git rebase -i` (and don't forget that git lets you define aliases!).
```
Comment (SuperSandro2000) [inline CONTRIBUTING.md:58]: IMO fixuping is an advanced concept. A far simpler workflow is to either do ``git ci -p --am`` or ``git ci`` and then rebase via ``git rebase --interactive @~3``.

**Comment by 7c6f434c** [comment]
It is obviously not a consensus _for all types of pull requests_.

Simple bumps need only one commit anyway, so squashing the fix-ups is the simplest approach.

Complicated layered changes are easier to review with chronologically structured commits (so that I can see that a part has not changed; maybe with something like evolving changesets it would work differently), and then if there is some in-repository documentation included maybe it matters less what the commit message says.

And if we really cared about detailed history, maybe we'd use a VCS that is capable of preserving distinction of what was done in staging and what in the mainline as first-class information.

**Pull request merged**